### PR TITLE
Report MSTest `[Retry]` attempts to the platform instead of discarding them

### LINKED ISSUE #10292: MSTest `[Retry]` attempts are discarded before they reach the platform, so in-process retries are invisible everywhere
## Summary

`RetryBaseAttribute.ExecuteAsync` returns a `RetryResult` holding **every** attempt, but the adapter keeps only the last one:

https://github.com/microsoft/testfx/blob/main/src/Adapter/MSTestAdapter.PlatformServices/Execution/UnitTestRunner.RunSingleTest.cs#L246

```csharp
result = retryResult.TryGetLast() ?? throw ApplicationStateGuard.Unreachable();
```

Every prior attempt is dropped there and never becomes a `TestNodeUpdateMessage`. As a result a test decorated with `[Retry(n)]` that fails and then passes is reported as an ordinary passing test — no attempt count, no flaky marker, in the terminal *or* in any report format.

The `RetryResult` XML doc already anticipates this:

> The other results are currently not used, but may be used in the future for tooling to show the state of the failed attempts.

This issue is that follow-up.

## Background and Motivation

Two concrete consequences, both verified against `main` with a `[Retry(3)]` test that fails on its first attempt and passes on its second:

**1. The terminal shows nothing.**

```
Test run summary: Passed! - RetryDemo.dll (net8.0|x64)
  total: 1
  failed: 0
  succeeded: 1
  skipped: 0
  duration: 378ms
```

**2. The CTRF report claims the test is not flaky.**

```json
{ "tests": 1, "passed": 1, "failed": 0, "skipped": 0, "flaky": 0 }
```

with no `retries` count and no `retryAttempts[]` array on the test entry.

That second point matters beyond cosmetics: `Microsoft.Testing.Extensions.CtrfReport` already implements the correct flaky rule in [`CtrfReportEngine.TestCollapsing.cs`](https://github.com/microsoft/testfx/blob/main/src/Platform/Microsoft.Testing.Extensions.CtrfReport/CtrfReportEngine.TestCollapsing.cs) (`CollapseAttempts` + `IsFlaky`), and `CtrfReportEngine.JsonTestWriter` already knows how to emit `retries` and `retryAttempts[]`. Because no prior attempt ever reaches the message bus, that entire code path is **unreachable for `[Retry]`** and the emitted report is silently wrong for consumers that key off `flaky`.

The out-of-process `--retry-failed-tests` orchestrator now reports flaky tests by name (#5362), which makes the asymmetry more visible: the same flakiness is surfaced when the whole host is relaunched, but hidden when `[Retry]` handles it in-process.

## Proposed Feature

Report each `[Retry]` attempt as a `TestNodeUpdateMessage` carrying its 1-based attempt number (a new `RetryAttemptProperty`, or an equivalent already-serialized property).

The consuming side largely exists already:

- [`TestProgressState.ReportGenericTestResult`](https://github.com/microsoft/testfx/blob/main/src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TestProgressState.cs) already implements "same uid, higher attempt → un-count the previous attempt, count it as retried, remember the earlier failure". The `flaky:` / `retried:` summary lines and the `Flaky tests:` section added in #5362 would therefore light up for `[Retry]` with no extra rendering work.
- CTRF's `CollapseAttempts` / `IsFlaky` would start producing correct `flaky`, `retries` and `retryAttempts[]` values.

### Things to work through

- **TRX/JUnit shape.** These formats would gain rows for prior attempts unless they go through the same per-uid collapse rule. This is the main risk and should be decided before implementing.
- **Attempt granularity for data-driven tests.** Attribution is per `TestNodeUid`, so a folded (`TestDataSourceUnfoldingStrategy.Fold`) data source reports one entry regardless of how many rows ran — the same documented limitation as #5362.
- **Interaction with `--retry-failed-tests`.** Attempt numbers are already used by the out-of-process orchestrator via `TESTINGPLATFORM_DOTNETTEST_ATTEMPTNUMBER`. When both are enabled the counts multiply (as `RetryAttribute`'s docs note), so the two attempt notions need to compose rather than collide.

## Alternative Designs

- **Report only a summary property** (e.g. "this test was retried N times, first outcome X") rather than a full message per attempt. Cheaper and avoids the TRX row-count question entirely, but loses per-attempt error messages and durations, and would not populate CTRF's `retryAttempts[]`.
- **Leave the message stream alone and expose the attempts through a side channel** consumed only by the terminal reporter. Avoids affecting report formats, but then the reports stay wrong, which is the more important half of the problem.
- **Do nothing and document it.** Rejected: the CTRF `flaky` field is machine-readable output that is currently incorrect rather than merely absent.

---

*Follow-up split out of #5362; the out-of-process half is implemented, this is the in-process half.*

### PR DESCRIPTION
**Pull request by Evangelink**
Implements #10292.

## Problem

`UnitTestRunner.RunSingleTestAsync` kept only the last retry attempt:

```csharp
result = retryResult.TryGetLast() ?? throw ApplicationStateGuard.Unreachable();
```

so an in-process `[Retry]` was invisible everywhere — the terminal showed a plain passing test, and the CTRF report claimed `flaky: 0` with no `retryAttempts[]` even though `CtrfReportEngine.CollapseAttempts` / `IsFlaky` already implement the correct rule.

## Approach

Publish **every** attempt as its own `TestNodeUpdateMessage` under the **same** `TestNode.Uid`, tagged with a new platform property:

```csharp
public sealed class RetryAttemptProperty : IProperty
{
    public RetryAttemptProperty(int attemptNumber, bool isSuperseded);
    public int AttemptNumber { get; }   // 1-based, within a single test-host run
    public bool IsSuperseded { get; }   // a later attempt for the same uid follows
}
```

`IsSuperseded` lets each consumer opt in or out without buffering the stream, via the shared `TestNode.IsSupersededRetryAttempt()` helper.

## Composes with #10329

#10329 ("Report retried and flaky tests by name") landed in main mid-review and reworked the same terminal subsystem. The two are complementary and now integrated — in-process `[Retry]` attempts feed the shared flaky reporting:

```
failed (try 1) FlakyTest (10s 000ms)
  Tests failed
passed (try 2) FlakyTest (10s 000ms)

Test run summary: Passed! - assembly.dll (net8.0|x64)
  total: 1
  failed: 0
  succeeded: 1
  skipped: 0
  flaky: 1 (passed after retry)
  retried: 1 test(s), 1 extra run(s)

Flaky tests:
  FlakyTest failed -> passed (2 attempts)
```

`TestProgressState` orders attempts by the pair `(host attempt, in-process attempt)`, and tracks the execution count per uid so the two mechanisms compose rather than collide. CTRF additionally emits `"flaky": 1`, `"retries": 1` and a populated `retryAttempts[]`.

## Design note: sibling property vs. part of the state property

Raised in review — should the attempt data hang off `Passed`/`Failed`/…? Two real points in favour: `PropertyBag` has a dedicated O(1) field for the state property, and enforces exactly one, which would make illegal states unrepresentable.

Kept as a sibling because the state types are sealed with public constructors and repo policy bans `init` on new public API, so carrying two values there means **~12–15 new public constructors** plus an equality change on shipped types. Attempt number is also the same category as `TimingProperty` (it describes the execution occurrence, not the outcome) and applies to `InProgress` too. The O(n) lookup is not a new class of cost — `TestApplicationResult` already does an O(n) `Any<TestNodeExecutionCompletedProperty>()` walk per node.

## Consumer audit

After several review rounds each surfaced another affected consumer, I swept every file referencing terminal state properties and classified each one.

**Filter superseded attempts** (want one row per test): `TestApplicationResult` (exit code), `AbortForMaxFailedTestsExtension`, `RetryDataConsumer`, `MSBuildConsumer`, `GitHubActionsAnnotationReporter`, `AzureDevOpsReporter`, `AzureDevOpsArtifactUploader`, `VideoRecorderSessionHandler`, `TrxDataConsumer`, `JUnitReportGenerator`.

**Keep the test tracked until the final attempt** (state machines, not tallies — a retry sequence has one in-progress update but a terminal update per attempt, so completing on a superseded one drops the test while it is still running): `HangDumpActivityIndicator`, `SimplifiedConsoleOutputDeviceBase`, `SlowTestReporterBase`.

**Deliberately unchanged:** `CtrfReportEngine` and `HtmlReportEngine` (keep the history by design — this is what fixes `flaky`), `CrashDumpSequenceLogger` (a diagnostic trace; the attempts leading to a crash are exactly what you want to see, so a guard would have been actively harmful), `PerRequestServerDataConsumerService` (already dedupes), `DotnetTestDataConsumer` (forwards by design).

VSTest is unaffected: its recorder drops superseded attempts, so Test Explorer and the VSTest TRX are byte-identical to before.

## Other correctness fixes found in review

- **Crash:** class/assembly cleanup results are appended after the retry sequence and kept attempt 1, so the terminal's attempt ordering saw an out-of-order arrival and threw `Unreachable()`. Cleanup results now inherit the attempt the test finished on.
- **Empty final attempt:** the combined array was non-empty (superseded only), so the empty-result/execution-error path was skipped, every entry was filtered out, and `AllPassed` vacuously returned `true` — turning a missing result into a *passing* test that unblocked dependents.
- **Flaky attempt count:** `Math.Max(host, retry)` undercounts when both mechanisms advance; now tracked per uid.
- **Server mode:** both JSON-RPC serializers silently drop unregistered properties, so attempt data vanished under `--server`. Registered in both (System.Text.Json + Jsonite), documented as `retry.attempt` / `retry.is-superseded`.
- **Serialization:** `TestResult` is `[Serializable]` on .NET Framework; the two new members needed `[field: OptionalField]` to match the existing convention for version-tolerant AppDomain deserialization.
- **BOM:** restored the UTF-8 BOM a scripted edit stripped, and added it to the new files per `.editorconfig`.

## Remaining follow-ups (not blockers)

- **`dotnet test` orchestrator path** does not carry the in-process attempt through the IPC wire model, so under `dotnet test` a `[Retry]` result is keyed `(host attempt, 1)` for every attempt. Needs a versioned wire-model field.

- **Progress renderer** advances once per attempt, so a retry-heavy run can exceed the discovered count. Verified pre-existing on `main` for host retries; the fix should cover both mechanisms at once.

- **Per-attempt timestamps are coarse.** `TestExecutionManager` measures `startTime`/`endTime` once around the whole `RunSingleTestAsync` call, so every flattened attempt carries the same window while keeping its own (correct) `Duration`. CTRF's `retryAttempts[]` therefore shows each attempt spanning the whole sequence — attempt 1's `stop` lands after attempt 2 began. Fixing it properly means capturing per-attempt boundaries, which needs new serializable `TestResult` members; simply omitting the timing would also drop the correct per-attempt duration.

- **Cleanup results inflate the retried-execution count.** A class/assembly cleanup failure is reported under the test's uid and is stamped with the final attempt number (so the attempt ordering does not see it as out-of-order). It then lands on `TestProgressState`'s same-attempt branch, which treats an extra result for an already-retried uid as another execution — so one retry plus a cleanup failure reports `2 extra run(s)` instead of `1`. That layer cannot distinguish a cleanup result from a data-driven row, so a fix needs a distinguishing signal through the three tracked-internal-API `Report*Test` overloads. Cosmetic and edge-case.

- **`outcome-kind`** is documented in the protocol (`"passed-with-retries"`) but implemented nowhere; now that attempts flow it could be populated for IDE filtering.

## Validation

- `Microsoft.Testing.Platform.UnitTests` 1983 · `Microsoft.Testing.Extensions.UnitTests` 879 · `MSTestAdapter.PlatformServices.UnitTests` 1071 · `TestFramework.UnitTests` 1502 (net9.0) and 1376 (net48, the `[Serializable]` path) — all passing
- Acceptance after `-pack`: MSTest retry/CTRF/JUnit/TRX and platform retry + max-failed-tests + MSBuild — all passing
- New behavioral tests were mutation-checked: disabling each guard fails its test

Fixes #10292

### CODE REVIEW

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TestProgressState.cs [hunk 1]
```diff
@@ -224,13 +226,22 @@ public int TryCount
...
 
     public void ReportFailedTest(string testNodeUid, string instanceId)
-        => ReportGenericTestResult(testNodeUid, instanceId, static entry => entry with { Failed = entry.Failed + 1 }, static @this => @this._failedTests++);
+        => ReportFailedTest(testNodeUid, instanceId, retryAttemptNumber: 1);
+
+    public void ReportPassingTest(string testNodeUid, string instanceId, int retryAttemptNumber)
+        => ReportGenericTestResult(testNodeUid, instanceId, retryAttemptNumber, static entry => entry with { Passed = entry.Passed + 1 }, static @this => @this._passedTests++);
+
+    public void ReportSkippedTest(string testNodeUid, string instanceId, int retryAttemptNumber)
+        => ReportGenericTestResult(testNodeUid, instanceId, retryAttemptNumber, static entry => entry with { Skipped = entry.Skipped + 1 }, static @this => @this._skippedTests++);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TestProgressState.cs]: Good catch — declared all the new signatures in `InternalAPI.Unshipped.txt` (the three `TestProgressState.Report*Test` overloads and the new `TerminalTestReporter.TestCompleted` overload), matching how the existing ones are tracked in `InternalAPI.Shipped.txt`.

One correction for the record: RS0016 did not actually fail the build — CI was green on the earlier head. The entries are still the right thing to add for consistency, so they are in.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: src/Adapter/MSTestAdapter.PlatformServices/Execution/UnitTestRunner.RunSingleTest.cs [hunk 2]
```diff
@@ -331,6 +331,58 @@ internal async Task<TestResult[]> RunSingleTestAsync(
...
+    /// <remarks>
+    /// Historically only the last attempt survived, which made an in-process retry invisible to every consumer -
+    /// the terminal, the reports, and notably CTRF's <c>flaky</c> flag (see
+    /// https://github.com/microsoft/testfx/issues/10292). The earlier attempts are kept here and tagged so each
+    /// recorder can decide what to do with them: the Microsoft.Testing.Platform recorder reports them all (tagged
+    /// with <c>RetryAttemptProperty</c>), while VSTest - which has no notion of attempts - drops the superseded
+    /// ones and keeps its historical single-result-per-test behavior.
+    /// <para>
+    /// The final attempt stays last so callers that look at <c>results[^1]</c> (class and assembly cleanup) and
+    /// consumers that collapse per test node uid keep observing the test's real outcome.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Adapter/MSTestAdapter.PlatformServices/Execution/UnitTestRunner.RunSingleTest.cs:348]: Confirmed and fixed. Class/assembly cleanup results are appended after the retry sequence and kept the default attempt number 1, so once a test finished on attempt 2+ the terminal''s attempt ordering saw an out-of-order arrival and threw `Unreachable()`.

Added `AlignAttemptNumber`, which stamps an appended cleanup result with the attempt the test actually finished on (walking back to the last non-superseded result). It is deliberately not marked superseded — a cleanup failure is part of the test''s final outcome. Applied at both append sites (class cleanup and assembly cleanup).

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: src/Adapter/MSTest.TestAdapter/TestingPlatformAdapter/MSTestTestNodeConverter.cs [hunk 3]
```diff
@@ -87,6 +87,14 @@ public static TestNode ToResultTestNode(UnitTestElement element, FrameworkTestRe
 
         testNode.Properties.Add(new TimingProperty(new(startTime, endTime, result.Duration), []));
 
+        // Surface an in-process retry (MSTest's [Retry]) so the platform can tell the attempts of one test apart
+        // instead of seeing repeated results for the same uid. Only added when a retry actually happened, so a
+        // regular test node is byte-identical to before.
+        if (result.RetryAttemptNumber > 1 || result.IsSupersededRetryAttempt)
+        {
+            testNode.Properties.Add(new RetryAttemptProperty(result.RetryAttemptNumber, result.IsSupersededRetryAttempt));
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Adapter/MSTest.TestAdapter/TestingPlatformAdapter/MSTestTestNodeConverter.cs:104]: You are right, and this was the most valuable comment in the review — my original consumer audit missed these entirely. All five now filter superseded attempts through the shared `TestNode.IsSupersededRetryAttempt()` helper:

- `AbortForMaxFailedTestsExtension` — `--maximum-failed-tests 1` could otherwise abort on a `[Retry]` test that goes on to pass.
- `RetryDataConsumer` — was requesting an out-of-process host relaunch for a test whose final in-process attempt passed.
- `MSBuildConsumer` / `GitHubActionsAnnotationReporter` — were logging build errors / annotations for a test that passes.
- `SimplifiedConsoleOutputDeviceBase` — was counting and printing superseded failures.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/TerminalOutputDevice.DataConsumption.cs [hunk 4]
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/TerminalOutputDevice.DataConsumption.cs]: Correct — a scripted edit of mine stripped it. I verified against `origin/main` that this file genuinely had a BOM before my change, so it was a real regression rather than pre-existing drift. Restored as UTF-8 with BOM.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: src/Platform/Microsoft.Testing.Platform/Messages/RetryAttemptProperty.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/Messages/RetryAttemptProperty.cs]: Fixed — saved as UTF-8 with BOM, along with the other new file added in this PR (`RetryAttemptPropertyExtensions.cs`).

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TestProgressState.cs [hunk 6]
```diff
@@ -371,13 +389,24 @@ private void ReportGenericTestResult(
...
             }
 
             incrementCountAction(this);
         }
     }
 
+    /// <summary>
+    /// Orders two attempts of the same test node. The out-of-process host attempt is the major component and the
+    /// in-process retry attempt the minor one, so the two retry mechanisms compose instead of colliding: host
+    /// attempt 2 / retry 1 is later than host attempt 1 / retry 5.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TestProgressState.cs:485]: Confirmed and fixed. `RetryDataConsumer` now ignores superseded attempts, so combining `[Retry]` with `--retry-failed-tests` no longer relaunches the host for a test whose final in-process attempt passed.

Worth noting the branch has since merged #10329, which reworked this same area — the two retry mechanisms now genuinely compose: in-process attempts feed the shared flaky/retried reporting, so a fail-then-pass `[Retry]` test is listed under `Flaky tests:` with its attempt count. Covered by a terminal reporter unit test asserting the full summary, and the existing `--retry-failed-tests` acceptance suite (32 tests) still passes.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: src/Platform/Microsoft.Testing.Platform/Services/TestApplicationResult.cs [hunk 7]
```diff
@@ -95,6 +95,15 @@ public Task ConsumeAsync(IDataProducer dataProducer, IData value, CancellationTo
             return Task.CompletedTask;
         }
 
+        // A test framework that retries a test in-process (MSTest's [Retry], ...) reports every attempt under the
+        // same test node uid. Only the final attempt is the test's outcome, so superseded attempts must not be
+        // counted here: otherwise a test that failed once and then passed would still leave _failedTestsCount > 0
+        // and make the process exit with ExitCode.AtLeastOneTestFailed.
+        if (message.TestNode.IsSupersededRetryAttempt())
+        {
+            return Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/Services/TestApplicationResult.cs:104]: Agreed — filtering the exit-code accumulator alone was not enough. `AbortForMaxFailedTestsExtension` now applies the same superseded-attempt guard, so `--maximum-failed-tests` no longer counts an attempt that a later one supersedes and cannot force exit code 13 on a fail-then-pass `[Retry]` test.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: src/Adapter/MSTestAdapter.PlatformServices/Execution/UnitTestRunner.RunSingleTest.cs [hunk 8]
```diff
@@ -331,6 +331,58 @@ internal async Task<TestResult[]> RunSingleTestAsync(
...
+    /// </para>
+    /// </remarks>
+    private static TestResult[] CombineRetryAttempts(TestResult[] firstAttempt, RetryResult retryResult)
+    {
+        IReadOnlyList<TestResult[]> retryAttempts = retryResult.AllResults;
+        if (retryAttempts.Count == 0)
+        {
+            // A RetryBaseAttribute implementation is expected to add at least one attempt.
+            throw ApplicationStateGuard.Unreachable();
+        }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Adapter/MSTestAdapter.PlatformServices/Execution/UnitTestRunner.RunSingleTest.cs:358]: Good catch — this was the subtlest issue found in the review. If the final attempt produced no results, the combined array was still non-empty (superseded entries only), so `SendTestResultsAsync` skipped its empty-result/execution-error path, every entry was then filtered out by the recorders, and `AllPassed` vacuously returned `true` — turning a missing result into a passing test that unblocked its dependents.

`CombineRetryAttempts` now returns an empty array when the final attempt produced no results, preserving the existing empty-result contract.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.TestCompletion.cs [hunk 9]
```diff
@@ -87,7 +127,12 @@ internal void TestCompleted(
             expected,
             actual,
             standardOutput,
-            errorOutput);
+            errorOutput,
+            // The orchestrator attributes retries per host instance; it does not surface a test framework's
+            // in-process retry attempt, so results arriving through this path are always attempt 1 of their host
+            // attempt.
+            retryAttemptNumber: 1,
+            isRetryAttempt: false);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.TestCompletion.cs]: Accurate diagnosis, and I am deliberately leaving this one open rather than fixing it here.

The analysis is right: under `dotnet test` the results travel as `SuccessfulTestResultMessage` / `FailedTestResultMessage`, neither of which has a retry field, so this overload can only pass attempt 1 and in-process retries are not distinguished on that path.

Fixing it properly means adding a field to the IPC wire model plus its serializer, the consumer, and this overload — a versioned protocol change touching a different subsystem from the rest of this PR, with its own compatibility considerations for older peers. That is more than this change should absorb, and doing it badly is worse than deferring it.

The in-process host path (the one this PR targets, and the one the issue is about) is fully wired. Leaving this thread unresolved so the protocol work stays visible for a human to schedule. Flagged in the PR description under remaining follow-ups.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.TestCompletion.cs [hunk 10]
```diff
@@ -128,13 +175,13 @@ private void TestCompleted(
...
-                asm.ReportFailedTest(testNodeUid, instanceId);
+                asm.ReportFailedTest(testNodeUid, instanceId, retryAttemptNumber);
                 break;
             case TestOutcome.Passed:
-                asm.ReportPassingTest(testNodeUid, instanceId);
+                asm.ReportPassingTest(testNodeUid, instanceId, retryAttemptNumber);
                 break;
             case TestOutcome.Skipped:
-                asm.ReportSkippedTest(testNodeUid, instanceId);
+                asm.ReportSkippedTest(testNodeUid, instanceId, retryAttemptNumber);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.TestCompletion.cs]: Valid observation, but this is pre-existing rather than introduced here, so I am leaving it open rather than expanding this PR.

`NotifyTestCompleted()` is already called unconditionally once per reported result on `main`, so the same over-count occurs today for `--retry-failed-tests`, where a host retry re-reports the same uid under a new instance id. This PR adds a new trigger for it; it does not create the behavior.

A correct fix means only advancing progress for a result that is not replacing a previous attempt for the same uid. That state lives in `TestProgressState.ReportGenericTestResult`, so surfacing it means changing the return type of the three `Report*Test` methods — all tracked internal API — and it should be fixed once for both retry mechanisms rather than bolted onto the in-process path alone.

Leaving this thread unresolved so the shared fix stays visible for a human. Flagged in the PR description under remaining follow-ups.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Platform/Microsoft.Testing.Platform/InternalAPI/InternalAPI.Unshipped.txt [hunk 11]
```diff
@@ -332,6 +334,9 @@ Microsoft.Testing.Platform.OutputDevice.Terminal.TerminalTestReporterOptions.Sho
 Microsoft.Testing.Platform.OutputDevice.Terminal.TerminalTestReporterOptions.ShowRunSummary.init -> void
 const Microsoft.Testing.Platform.OutputDevice.Terminal.TerminalTestReporterCommandLineOptionsProvider.ShowFlakyTestsOption = "show-flaky-tests" -> string!
 static Microsoft.Testing.Platform.OutputDevice.Terminal.TerminalTestReporterCommandLineOptionsProvider.IsFlakyTestsReportingEnabled(Microsoft.Testing.Platform.CommandLine.ICommandLineOptions! commandLineOptions) -> bool
+Microsoft.Testing.Platform.OutputDevice.Terminal.TestProgressState.ReportPassingTest(string! testNodeUid, string! displayName, string! instanceId, int retryAttemptNumber) -> void
+Microsoft.Testing.Platform.OutputDevice.Terminal.TestProgressState.ReportSkippedTest(string! testNodeUid, string! displayName, string! instanceId, int retryAttemptNumber) -> void
+Microsoft.Testing.Platform.OutputDevice.Terminal.TestProgressState.ReportFailedTest(string! testNodeUid, string! displayName, string! instanceId, int retryAttemptNumber) -> void
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/InternalAPI/InternalAPI.Unshipped.txt:340]: Verified and fixed — you are right. I had declared the three `TestProgressState.Report*Test` overloads but missed the new 14-parameter `TerminalTestReporter.TestCompleted` one; the existing entry nearby turned out to be `TestCompletedWithoutResult`, not this overload, so it was genuinely undeclared.

Added it alongside the `TestProgressState` entries, copying the parameter formatting from the 12-parameter overload in `InternalAPI.Shipped.txt` so the nullability annotations match. The build passing confirms the declaration matches the real signature (a mismatch would trip RS0017).

One clarification carried over from the earlier thread on the same topic: this is a consistency fix, not a build fix — RS0016 was not failing CI, which was green on the previous head.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

**Review by 0101 — APPROVED** [review]

**Review by 0101 — APPROVED** [review]

**Review by 0101 — APPROVED** [review]

File: src/TestFramework/TestFramework/Attributes/TestMethod/TestResult.cs [hunk 12]
```diff
@@ -215,6 +215,24 @@ public Exception? TestFailureException
...
+    /// <summary>
+    /// Gets or sets the 1-based attempt this result belongs to when the test method is decorated with a
+    /// <see cref="RetryBaseAttribute"/>. The first (non-retry) execution is attempt 1.
+    /// </summary>
+    /// <remarks>
+    /// Every attempt is reported to the test host so tooling can surface the in-process retry (see the
+    /// <c>RetryAttemptProperty</c> platform property). Results that are not part of a retry sequence keep the
+    /// default value of 1, which reports as "no retry happened" everywhere.
+    /// </remarks>
+    internal int RetryAttemptNumber { get; set; } = 1;
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Attributes/TestMethod/TestResult.cs:237]: Verified and fixed. `TestResult` is `[Serializable]` under `#if NETFRAMEWORK` because it crosses AppDomain boundaries, and the previously added members (`ExceptionExpectedText` / `ExceptionActualText`) already carry `[field: OptionalField]` guarded the same way — so this is an established convention in the file that the two retry members were missing. Applied it to both.

One detail worth recording, since it differs slightly from the string members you cited: `RetryAttemptNumber` has an initializer of `1`, and field initializers do not run during deserialization, so an older payload yields `0` rather than `1`. That is harmless here — every consumer keys off `IsSupersededRetryAttempt` (which defaults to `false`, i.e. "this is the final outcome"), and the attempt number is only ever used for display once a retry is known to have happened. I documented that on the member rather than contriving a fake `1`.

Built and ran the .NET Framework target specifically, since it is the only one where this code path exists: 1376 tests pass on net48.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

**Review by 0101 — APPROVED** [review]

**Review by 0101 — APPROVED** [review]
